homeassistant/components/bthome/event.py

```diff
@@ -0,0 +1,151 @@
+        self.entity_description = replace(base_description, key=event_class)
+        postfix_one_indexed = int(postfix) + 1 if postfix else None
+        postfix_name = f" {postfix_one_indexed}" if postfix else ""
+        # We report button button 1 as button 2, etc to the user since
+        # the first button has no postfix, and it would be confusing
+        # to report the second button as button 1
+        self._attr_name = f"{base_event_class.title()}{postfix_name}"
+        # Matches logic in PassiveBluetoothProcessorEntity
+        if device_id:
+            self._attr_device_info = dr.DeviceInfo(
+                identifiers={(DOMAIN, f"{address}-{device_id}")}
+            )
+            self._attr_unique_id = f"{address}-{event_class}-{device_id}"
+        else:
+            self._attr_device_info = dr.DeviceInfo(
+                identifiers={(DOMAIN, address)},
+                connections={(dr.CONNECTION_BLUETOOTH, address)},
+            )
+            self._attr_unique_id = f"{address}-{event_class}"
+
+    async def async_added_to_hass(self) -> None:
+        """Entity added to hass."""
+        await super().async_added_to_hass()
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                self._update_signal,
+                self._handle_event,
+            )
+        )
+
+    @callback
+    def _handle_event(self, event: BTHomeBleEvent) -> None:
+        self._trigger_event(event[EVENT_TYPE], event[EVENT_PROPERTIES])
+        self.async_write_ha_state()
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up BTHome event."""
+    coordinator: BTHomePassiveBluetoothProcessorCoordinator = hass.data[DOMAIN][
+        entry.entry_id
+    ]
+    address = coordinator.address
+    discovered_device_keys: set[DeviceKey] = set()
+    ent_reg = er.async_get(hass)
+    to_add: list[BTHomeEventEntity] = []
+    for ent_reg_entry in er.async_entries_for_config_entry(ent_reg, entry.entry_id):
+        if ent_reg_entry.domain != "event":
+            continue
+        unique_id_split = ent_reg_entry.unique_id.split("-")
+        device_id: str | None = None
+        # Matches logic in PassiveBluetoothProcessorEntity
+        if len(unique_id_split) == 3:
+            address, event_class, device_id = unique_id_split
+        else:
+            address, event_class = unique_id_split
+            device_id = None
+        discovery_key = DeviceKey(event_class, device_id)
+        discovered_device_keys.add(discovery_key)
+        to_add.append(BTHomeEventEntity(address, event_class, device_id))
+
+    async_add_entities(to_add)
+
+    @callback
+    def _async_discovered_device_key(device_key: DeviceKey) -> None:
+        """Handle a discovered device key."""
+        if device_key in discovered_device_keys:
+            return
+        discovered_device_keys.add(discovery_key)
```

Comment:
```suggestion
        discovery_key = DeviceKey(event_class, device_id)
        discovered_device_keys.add(discovery_key)
```

I got an error on this line when testing a shelly button. Fixed it by replacing this line. The error was

```
2024-01-20 12:51:29.666 ERROR (MainThread) [homeassistant.util.logging] Exception in _async_discovered_device_key when dispatching 'bthome_discovered_device_key_B4:35:22:F5:3D:A4': (DeviceKey(key='button', device_id=None),)
Traceback (most recent call last):
  File "/home/pi/core/homeassistant/components/bthome/event.py", line 140, in _async_discovered_device_key
    discovered_device_keys.add(discovery_key)
                               ^^^^^^^^^^^^^
NameError: cannot access free variable 'discovery_key' where it is not associated with a value in enclosing scope
```